src/track/globaltrackcache.cpp

```diff
@@ -30,240 +36,354 @@ TrackRef createTrackRef(const Track& track) {
+    if (m_trackRef.getId().isValid()) {
+        // Ignore initializing the same id twice
+        DEBUG_ASSERT(m_trackRef.getId() == trackId);
+    } else {
+        m_trackRef = m_pInstance->initTrackId(
+                m_strongPtr,
+                m_trackRef,
+                trackId);
+    }
+    unlockCache();
+    DEBUG_ASSERT(m_trackRef == createTrackRef(*m_strongPtr));
 }
 
-
 //static
-GlobalTrackCache* volatile GlobalTrackCache::s_pInstance = nullptr;
-
-//static
-void GlobalTrackCache::createInstance(GlobalTrackCacheEvictor* pEvictor) {
-    DEBUG_ASSERT(s_pInstance == nullptr);
-    s_pInstance = new GlobalTrackCache(pEvictor);
+void GlobalTrackCache::createInstance(GlobalTrackCacheDeleter* pDeleter) {
+    DEBUG_ASSERT(!s_pInstance);
+    s_pInstance = new GlobalTrackCache(pDeleter);
 }
 
 //static
 void GlobalTrackCache::destroyInstance() {
-    DEBUG_ASSERT(s_pInstance != nullptr);
+    DEBUG_ASSERT(s_pInstance);
     GlobalTrackCache* pInstance = s_pInstance;
+    // Reset the static/global pointer before entering the destructor
     s_pInstance = nullptr;
+    // Delete the singular instance
     delete pInstance;
 }
 
+void GlobalTrackCache::deleteTrack(Track* plainPtr) {
+    VERIFY_OR_DEBUG_ASSERT(plainPtr) {
+        kLogger.warning()
+                << "Cannot delete null track pointer";
+        return;
+    }
+    if (traceLogEnabled()) {
+        plainPtr->dumpObjectInfo();
+    }
+    // Track object must not be deleted by operator new!
+    // Otherwise the deleted track object might be accessed
+    // when processing cross-thread signals that are delayed
+    // within a queued connection and may arrive after the
+    // object has already been deleted.
+    if (debugLogEnabled()) {
+        kLogger.debug()
+                << "Deleting track"
+                << plainPtr;
+    }
+    plainPtr->deleteLater();
+}
+
 //static
-void GlobalTrackCache::deleter(Track* pTrack) {
-    if (s_pInstance != nullptr) {
-        s_pInstance->evict(pTrack);
+void GlobalTrackCache::deleter(Track* plainPtr) {
+    DEBUG_ASSERT(plainPtr);
+    // Any access to plainPtr is forbidden!! Due to race condition
+    // this pointer might already have been deleted! This happens
+    // when a previous invocation is outpaced by a following
+    // invocation that will already delete the object behind the
+    // pointer. Tracks might be revived before being deleted and
+    // as a consequence the reference count might drop to 0 multiple
+    // times. The order in which those competing invocations finally
+    // lock and enter the thread-safe cache is undefined!!!
+    if (s_pInstance) {
+        s_pInstance->evictAndDelete(plainPtr);
+    } else {
+        // Simply delete unreferenced tracks when the cache is no
+        // longer available. This might but should not happen.
+        kLogger.warning()
+                << "Deleting uncached track";
+        deleteTrack(plainPtr);
```

Comment:
This can be crashing if the issue above happens. 
I still suffer a shut don crash rarely.Could this be the issue?

Reply:
This actually could be an issue if we would destroy the cache while multiple deletions for the same track object are pending. I agree, we should just print a _critical_ log message and don't touch the pointer at all.

But since we decided not to destroy the singleton while shutting down this can never happen! Did you see the warning in the log?

Reply:
No, I did not see the warning. Not sure, but it can be hidden by the delayed flushing?